src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,523 @@
+            // 1) We guessed exactly how many threads there are
+            // 2) There are more threads that we didn't get since our buffer is too small
+            // To make sure it isn't #2, when we result == size, double the buffer and try again
+            while (result == size)
+            {
+                Marshal.FreeCoTaskMem(buffer);
+                size *= 2;
+                Marshal.AllocCoTaskMem(size);
+                result = proc_pidinfo(pid, PROC_PIDLISTTHREADS, 0, buffer, size);
+            }
+
+            // Any data means result > 0
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            List<KeyValuePair<ulong, proc_threadinfo?>> threads = new List<KeyValuePair<ulong, proc_threadinfo?>>();
+
+            // Loop over each thread and get the info
+            for (int i = 0; i < (result / PROC_PIDLISTTHREADS_SIZE); i++)
+            {
+                ulong threadId = (ulong)Marshal.ReadInt64(buffer, i * Marshal.SizeOf<ulong>());
+                threads.Add(new KeyValuePair<ulong, proc_threadinfo?>(threadId, GetThreadInfoById(threadId)));
+            }
+
+            // Free our unmanaged memory
+            Marshal.FreeCoTaskMem(buffer);
+
+            return threads;
+        }
+
+        /// <summary>
+        /// Retrieves the number of open file descriptors for the specified pid
+        /// </summary>
+        /// <returns>Returns a list of open File Descriptors for this process</returns>
+        /// <remarks>
+        /// This function doesn't use the helper since it seems to allow passing NULL
+        /// values in to the buffer and length parameters to get back an estimation 
+        /// of how much data we will need to allocate; the other flavors don't seem
+        /// to support doing that.
+        /// </remarks>
+        internal static List<proc_fdinfo> GetFileDescriptorsForPid(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int size = 0;
+            IntPtr buffer = IntPtr.Zero;
+            
+            // Query for an estimation about the size of the buffer we will need. This seems
+            // to add some padding from the real number, so we don't need to do that
+            int result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, buffer, size);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            // Allocate our unmanaged memory and call again to get the descriptors
+            size = result;
+            buffer = Marshal.AllocCoTaskMem(size);
+            result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, buffer, size);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
```

Comment:
If we throw here, we leak the memory in buffer.  You should use try { } finally to ensure we release the memory in case of an exception.